### PR TITLE
[FLINK-8312][TableAPI && SQL] Fix ScalarFunction varargs length excee…

### PR DESCRIPTION
**Pull request by Xpray**
## What is the purpose of the change

*Support parameters exceeds 254 with Varargs for SQL*

## Brief change log

  - *if scalar function is varargs, do not check parameter length, set max length to -1*

## Verifying this change

This change added tests and can be verified as follows:

*(example:)*
  - *SqlITCase.testUDFWithLongVarargs*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/SqlITCase.scala [hunk 1]
```diff
@@ -481,4 +484,34 @@ class SqlITCase extends StreamingWithStateTestBase {
...
+    val expected = List(
+      "510",
+      "1275",
+      "2805")
+    assertEquals(expected.sorted, StreamITCase.testResults.sorted)
+  }
+
+}
+
+object VarUDF extends ScalarFunction {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/SqlITCase.scala]: Suggest using exist test scalarFunction, such as : 
`org.apache.flink.table.expressions.utils.userDefinedScalarFunctions # Func15`
OR move the VarUDF into `org.apache.flink.table.expressions.utils.userDefinedScalarFunctions`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/ScalarSqlFunction.scala [hunk 2]
```diff
@@ -153,15 +154,19 @@ object ScalarSqlFunction {
       override def getOperandCountRange: SqlOperandCountRange = {
         var min = 255
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/ScalarSqlFunction.scala]: min=254 is enough.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/ScalarSqlFunction.scala [hunk 3]
```diff
@@ -153,15 +154,19 @@ object ScalarSqlFunction {
...
+        var isVarargs = false
         signatures.foreach( sig => {
-          var len = sig.length
-          if (len > 0 && sig(sig.length - 1).isArray) {
-            max = 254  // according to JVM spec 4.3.3
-            len = sig.length - 1
+          var len = sig._2.length
+          if (len > 0 && sig._1 && sig._2(sig._2.length - 1).isArray) {
+            isVarargs = true
+            len = sig._2.length - 1
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/ScalarSqlFunction.scala]: **approach 1：**
sig._2.length - 1 => len -1 
len = sig._2.length - 1 => len-1

**approach2**
methods.foreach(
          m => {
            var len = m.getParameterTypes.length
            if (len > 0 && m.isVarArgs && m.getParameterTypes()(len - 1).isArray) {
              isVarargs = true
              len = len - 1
            }
            max = Math.max(len, max)
            min = Math.min(len, min)
          })

Using approach2 we can remove 
"val signatures = methods.map(m => m.isVarArgs -> m.getParameterTypes)"

What do you think?

**Comment by Xpray** [comment]
Hi, @sunjincheng121 , thanks for the comments, I've updated this PR.

**Comment by sunjincheng121** [comment]
@Xpray  Thanks for the update.
+1 to merged.

Best, Jincheng

**Pull request closed without merging**